### PR TITLE
[java-micronaut] Generate visitor for subtypes with a discriminator

### PR DESCRIPTION
**Pull request by auke-**
When types which extend a common type and are distinguished based on a
discriminator are consumed they are often cast to their specific Java
type which results in error prone boilerplate code.

By generating a visitor for those kind of types the various subtypes can
be consumed in a type safe manner.

This change is based on the same implementation I made for the [Kokuwa.io](https://github.com/kokuwaio/micronaut-openapi-codegen/pull/66)
version of the micronaut openapi plugin.

Note that there is also a pull request for the [typescript-angular](https://github.com/OpenAPITools/openapi-generator/pull/9278) implementation.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@andriy-dmytruk

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/java-micronaut/common/model/pojo.mustache [hunk 1]
```diff
@@ -355,4 +355,36 @@ Declare the class with extends and implements
...
+    public abstract <T> T accept(Visitor<T> visitor);
+
+    /**
+     * A {{classname}} visitor implementation allows visiting the various {{classname}} types.
+     *
+     * @param <R> the return type of the visitor
+     */
+    public interface Visitor<R> {
+        {{#discriminator.mappedModels}}
+        public R visit{{modelName}}({{modelName}} value);
```
Comment (andriy-dmytruk) [inline modules/openapi-generator/src/main/resources/java-micronaut/common/model/pojo.mustache:376]: The `public` modifier is redundant inside an interface, but it is Ok.

**Comment by andriy-dmytruk** [comment]
@auke- Everything looks good to me. Nice pattern and implementation. Thanks.

I am wondering if it could make sense to provide a default visitor implementation. Something like:
```java
    public interface DefaultVisitor<R> extends Visitor<R> {
        {{#discriminator.mappedModels}}
        default R visit{{modelName}}({{modelName}} value) {
            return visitDefault(value);
        }
        {{/discriminator.mappedModels}}

        R visitDefault({{classname}} value);
    }
```
which could be used when someone needs functionality analogous to `switch(...) default: ... `. But I suppose it can be arguable, as this is not exactly following the pattern. What do you think?

**Comment by auke-** [comment]
> I am wondering if it could make sense to provide a default visitor implementation.

For me a default visitor implementation is an anti-pattern since it prevents the compiler from warning me when an implementation is incomplete (either with an initial implementation or when the visitor interface is changed).

But since your proposal is an opt-in for the user, they don't have to use the DefaultVisitor and can choose if they want the fallback behavior or not. But i'm not sure if anyone would really want that.

**Comment by andriy-dmytruk** [comment]
> For me a default visitor implementation is an anti-pattern since it prevents the compiler from warning me when an implementation is incomplete (either with an initial implementation or when the visitor interface is changed).

Makes sense. And if anyone would really need that, they can create a feature request.

@wing328 can you merge this PR?

**Pull request merged**